controllers/signer/interface.go

```diff
@@ -18,10 +18,33 @@ package signer
 
 import (
 	"context"
+	"crypto/x509"
 
 	cmapi "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cert-manager/issuer-lib/api/v1alpha1"
 )
 
-type Sign func(ctx context.Context, cr *cmapi.CertificateRequest, issuerObject client.Object) ([]byte, error)
-type Check func(ctx context.Context, issuerObject client.Object) error
+type Sign func(ctx context.Context, cr CertificateRequestObject, issuerObject v1alpha1.Issuer) ([]byte, error)
+type Check func(ctx context.Context, issuerObject v1alpha1.Issuer) error
+
+// CertificateRequestObject is an interface that represents either a
+// cert-manager CertificateRequest or a Kubernetes CertificateSigningRequest
+// resource. This interface hides the spec fields of the underlying resource
+// and exposes a Certificate template and the raw CSR bytes instead. This
+// allows the signer to be agnostic of the underlying resource type and also
+// agnostic of the way the spec fields should be interpreted, such as the
+// defaulting logic that is applied to it. It is still possible to access the
+// labels and annotations of the underlying resource or any other metadata
+// fields that might be useful to the signer. Also, the signer can use the
+// GetConditions method to retrieve the conditions of the underlying resource.
+// To update the conditions, the signer should return an appropriate error
+// from the Sign method.
+type CertificateRequestObject interface {
+	metav1.Object
+
+	GetRequest() (template *x509.Certificate, csr []byte, err error)
```

Comment:
NOTE: I identified that the "duration" value is still missing here (since it cannot be easily determined by looking at the x509.Certificate object.

Reply:
fixed in https://github.com/cert-manager/issuer-lib/pull/20/commits/2427784b3a41157a37a92ca750c7a8242ddac3f0